Source/Core/Resource.js

```diff
@@ -0,0 +1,534 @@
+        }
+        if (defined(options.method)) {
+            resource.method = options.method;
+        }
+        if (defined(options.data)) {
+            resource.data = options.data;
+        }
+        if (defined(options.overrideMimeType)) {
+            resource.overrideMimeType = options.overrideMimeType;
+        }
+        if (defined(options.proxy)) {
+            resource.proxy = options.proxy;
+        }
+        if (defined(options.request)) {
+            resource.request = options.request;
+        }
+        if (defined(options.isDirectory)) {
+            resource._isDirectory = options.isDirectory;
+        }
+        if (defined(options.retryCallback)) {
+            resource.retryCallback = options.retryCallback;
+        }
+        if (defined(options.retryAttempts)) {
+            resource.retryAttempts = options.retryAttempts;
+        }
+
+        return resource;
+    };
+
+    /**
+     * Called when a resource fails to load. This will call the retryCallback function if defined until retryAttempts is reached.
+     *
+     * @param {Error} [error] The error that was encountered.
+     *
+     * @returns {Promise<Boolean>} A promise to a boolean, that if true will cause the resource request to be retried.
+     */
+    Resource.prototype.retryOnError = function(error) {
+        var retryCallback = this.retryCallback;
+        if ((typeof retryCallback !== 'function') || (this._retryCount >= this.retryAttempts)) {
+            return when(false);
+        }
+
+        var that = this;
+        return when(retryCallback(this, error))
+            .then(function(result) {
+                ++that._retryCount;
+
+                return result;
+            });
+    };
+
+    /**
+     * Duplicates a Resource instance.
+     *
+     * @param {Resource} [result] The object onto which to store the result.
+     *
+     * @returns {Resource} The modified result parameter or a new Resource instance if one was not provided.
+     */
+    Resource.prototype.clone = function(result) {
+        if (!defined(result)) {
+            result = new Resource({
+                url : this._url
+            });
+        }
+
+        result._url = this._url;
+        result._queryParameters = this._queryParameters;
+        result._templateValues = this._templateValues;
+        result.headers = this.headers;
+        result.request = this.request;
+        result.responseType = this.responseType;
+        result.method = this.method;
+        result.data = this.data;
+        result.overrideMimeType = this.overrideMimeType;
+        result.proxy = this.proxy;
+        result._isDirectory = this._isDirectory;
+        result.retryCallback = this.retryCallback;
+        result.retryAttempts = this.retryAttempts;
+        result._retryCount = 0;
+
+        return result;
+    };
+
+    /**
+     * Called when a request succeeds
+     */
+    Resource.prototype.succeeded = function() {
```

Comment:
What's the purpose of this function (outside the obvious).  I can see weird situation where you are using the same resources in two places and one succeeds and one fails and weird stuff happens.  Sure it's an edge case but it's a result of the Resource representing success/failure state without the Resource itself being the one that makes the request.

One possible option I can see is "binding" a particular resource to a particular action so that the resource itself can't be used simultaneously (but derived resources can still be created).  Basically, if I call `loadJson`, it checks that a flag on the provided resource is false and sets it to true until `loadJson` is done with it.  If `loadJson` detects the flag is already true, it through.  Maybe I'm just being paranoid. (hopefully I'm explaining the problem clearly). CC @shunter

Reply:
Low level loading functions (`loadWithXhr`, `loadImage` and `loadJsonp`) now clone the Resource properly, so this should no longer be an issue. Now that I think about it, this function can now just go away, because no one can reuse the cloned `Resource` anyway.